planner/core/preprocess.go

```diff
@@ -1594,10 +1594,23 @@ func (p *preprocessor) handleAsOfAndReadTS(node *ast.AsOfClause) {
 	}
 	if p.LastSnapshotTS != 0 {
 		dom := domain.GetDomain(p.ctx)
-		p.InfoSchema, p.err = dom.GetSnapshotInfoSchema(p.LastSnapshotTS)
-		if p.err != nil {
+		is, err := dom.GetSnapshotInfoSchema(p.LastSnapshotTS)
+		// if infoschema is empty, LastSnapshotTS init failed
+		if err != nil || is == nil {
+			p.err = err
```

Comment:
If you think `GetSnapshotInfoSchema` will return a nil inforschema without any error. You should make a standalone if statement:

```
if err != nil {
  p.err = err
  return
}
if is == nil {
  p.err = errors.New("your error")
  return 
}
```